apps/opt.c

```diff
@@ -656,22 +659,54 @@ int opt_next(void)
             /* Just a string. */
             break;
         case '/':
-            if (app_isdir(arg) >= 0)
+            if (app_isdir(arg) > 0)
                 break;
             BIO_printf(bio_err, "%s: Not a directory: %s\n", prog, arg);
             return -1;
         case '<':
             /* Input file. */
-            if (strcmp(arg, "-") == 0 || app_access(arg, R_OK) >= 0)
+            if (strcmp(arg, "-") == 0 || app_access(arg, R_OK) == 0)
                 break;
             BIO_printf(bio_err,
                        "%s: Cannot open input file %s, %s\n",
                        prog, arg, strerror(errno));
             return -1;
         case '>':
             /* Output file. */
-            if (strcmp(arg, "-") == 0 || app_access(arg, W_OK) >= 0 || errno == ENOENT)
+            l = strlen(arg);
+            if (arg[l - 1] == '/' || app_isdir(arg) > 0) {
```

Comment:
> Hmmm... it would seem to me like checking app_isdir(app_dirname(arg)) followed by app_access(arg) would be right then? I.e. make sure the parent is an existing directory before checking the writability of the output file itself, yeah?

Almost yes, one missing stuff is `arg` itself must not be a directory. But we cannot simply use `!app_isdir(arg)` to judge this, since if `arg` is something like `/tmp/intended-non-existent-file` will also be judged with 'not a directory' since it's not exist yet, but this is a *valid* case we need to further check.

So checking errno set by `app_isdir` while it's return-value is not zero would be a possible way to solve this.